src/prefect/engine.py

```diff
@@ -618,7 +618,14 @@ async def orchestrate_flow_run(
         )
 
     if flow.flow_run_name:
-        flow_run_name = flow.flow_run_name.format(**parameters)
+        if callable(flow.flow_run_name):
+            flow_run_name = flow.flow_run_name(parameters)
+        elif isinstance(flow.flow_run_name, str):
+            flow_run_name = flow.flow_run_name.format(**parameters)
+        else:
+            raise RuntimeError(
+                "flow_run_name parameter is neither a str nor a callable"
+            )
```

Comment:
In both `flow_run_name` and `task_run_name` I'm raising exceptions when they are neither `str` nor `callable`. Should this assertion be removed? If not should an assertion be added in the actual `Flow` and `Task` classes too?

Reply:
Raising errors at the flow/task class level is better — then we will help people do the right thing when writing their task / flow rather than at runtime.

Reply:
I'll make changes to raise `TypeError`'s in `Flow` & `Task` class level. I'm also thinking to leave the `RuntimeError`'s here. Please confirm if the same if acceptable.

Reply:
It's okay here as well, but we'll probably want to move this into a separate function since it's kind of long now.